### ISSUE TITLE
CoreML EP: allow MLProgram model format + add XNNPACK provider (NeuralNetwork default aborts layout inference on Apple Silicon)

### ISSUE THREAD

**Issue by petri** [docling-project/docling.rs#324]
## Summary

`docling-pdf` registers the CoreML execution provider with `ort::ep::CoreML::default().build()` (`src/ep.rs`, `dispatches()`) and offers no way to configure it. `CoreML::default()` sets no options, so ONNX Runtime's C-API defaults apply: **`ModelFormat=NeuralNetwork`**, `MLComputeUnits=ALL`, dynamic input shapes allowed.

With the bundled `layout_heron.onnx`, the CoreML `NeuralNetwork` backend takes graph partitions at session-creation time that it then cannot execute, and ONNX Runtime aborts inference — there is no per-node runtime fallback to CPU, so the whole PDF conversion fails.

```
[E:onnxruntime:, sequential_executor.cc:671 ExecuteKernel] Non-zero status code returned while running
  ..._CoreML_..._37 node. Status Message: Error executing model: Unable to compute the prediction using
  a neural network model. It can be an invalid input data or broken/unsupported model (error code: -1).
docling-pdf: layout model rejected a 2-page batch (...); falling back to per-page inference
<pdf>: failed (conversion failed: pdf: page 1: layout: inference: Non-zero status code returned ...)
```

Per-page inference fails the same way, so this is not the batch-shape path.

### Environment

- macOS 26 (Darwin 25.6.0), Apple Silicon
- `docling` 1.1.1, features `pdf` + `coreml`
- `ort` 2.0.0-rc.13

### Why `NeuralNetwork` is the wrong default here

`models/layout_heron.onnx` (opset 17):

- input `pixel_values`: `[batch, 3, 640, 640]` — dynamic batch
- outputs `logits` / `pred_boxes`: fully dynamic (`[batch, <dyn>, 17]`, `[batch, <dyn>, <dyn>]`)
- op set includes `GridSample`, `ScatterND`, `GatherElements`, `TopK`, `LayerNormalization`, `Range`, `Where`

The CoreML `NeuralNetwork` format has thin/absent support for several of these (notably `GridSample`) and copes poorly with dynamic output shapes. The newer `MLProgram` format covers far more operators — `ort`'s own docs note it "supports more operators, and may be more performant". This is the same `error code: -1` / "broken/unsupported model" signature widely reported for ORT + CoreML `NeuralNetwork`:

- https://github.com/microsoft/onnxruntime/issues/26355
- https://github.com/immich-app/immich/issues/25058
- https://github.com/chroma-core/chroma/issues/2555

`DOCLING_RS_EP=auto` does not help: it still registers CoreML ahead of CPU, and the failure is at execution, not registration.

## Requests

1. **Expose CoreML EP options**, at least `ModelFormat` and `MLComputeUnits`. The pinned `ort` 2.0.0-rc.13 already provides `CoreML::default().with_model_format(ep::coreml::ModelFormat::MLProgram)`, `.with_compute_units(...)`, and `.with_static_input_shapes(bool)`. An env var alongside `DOCLING_RS_EP` (e.g. `DOCLING_RS_COREML_MODEL_FORMAT`, `DOCLING_RS_COREML_COMPUTE_UNITS`) would match the existing selection model in `ep.rs`. Consider defaulting `coreml` / `auto` to `MLProgram` on macOS 12+.

2. **Add an XNNPACK provider** (`ort::ep::XNNPACK`, available in rc.13). ARM NEON / SIMD CPU acceleration is a solid speedup over the plain CPU provider for the `Conv` / `MatMul` / `Gemm`-heavy layout and OCR models, is cross-platform, needs no GPU, and sidesteps CoreML's op-coverage problem — a good default accelerator for Apple Silicon while CoreML support matures.

3. *Optional:* under `auto`, if a CoreML session errors at inference, retry that one session CPU-only instead of failing the document. (The `auto` branch currently builds CoreML without `.error_on_failure()`, but that only covers registration, not this runtime abort.)

## Workaround for downstream users

Set `DOCLING_RS_EP=cpu`. `auto` is unsafe as a default whenever the `coreml` feature is compiled in.

**Comment by artiz** [comment]
@petri please test fixed version from https://github.com/artiz/docling.rs/tree/claude/coreml-ep-324
If issue was resolved the I will merge it.


I use my personal fork to create PRs here because of Claude integration issues.

**Comment by petri** [comment]
Tested on an M4 Max (12P+4E CPU, 40-core GPU), macOS 26, docling.rs 1.31, against `docling-layout-heron`. Thanks for turning this around so quickly — the knobs are useful and `DOCLING_RS_COREML_STATIC_SHAPES` in particular is load-bearing. But **I'd suggest not merging the `MLProgram` default as-is**. Findings below, with what I could and couldn't control for.

## 1. `MLProgram` alone doesn't fix it, and changes how it fails

With the stock (dynamic-batch) `layout_heron.onnx`, `DOCLING_RS_EP=coreml` (and `auto`) **aborts the process** rather than returning an error:

```
mps.matmul op contracting dimensions differ 1 & 256
MPSGraphExecutable.mm:1484: failed assertion `original module failed verification'
→ SIGABRT (exit 134)
```

Previously this was a catchable `error code: -1` that failed one document. An uncatchable `abort()` is worse for callers, and `auto` is the default whenever the `coreml` feature is compiled in. I only observed this on one machine/OS, so treat the specific MPS assertion as anecdotal — but the change from "error" to "abort" seems worth checking before it becomes the default.

`DOCLING_RS_COREML_STATIC_SHAPES=1` avoids it. That knob works.

## 2. The larger lever looks like the model export, not the EP config

Re-exporting the layout model with a static batch (dropping `dynamic_axes`) changes CoreML's partitioning substantially:

| model | nodes taken by CoreML | partitions |
|---|---|---|
| stock (dynamic batch) | 1685 / 2626 | 61 |
| static batch=1 | 958 / 970 | 8 |

(Different node totals because the static export folds more constants — the two rows aren't directly comparable as ratios, but the partition-count difference is the point.)

Worth noting `scripts/install/export_layout.py` already builds this static graph internally as its numeric reference before discarding it.

## 3. Two cases where output is wrong without any error

Both measured against the same model on the CPU EP as reference, same input image:

- **`MLComputeUnits` defaulting to `ALL`** can select the Neural Engine, which runs fp16: `max|Δlogits| = 6.5`, `max|Δboxes| = 0.945` (boxes normalised 0–1). No error is raised. It was also slower than the GPU path here (113 vs 41 ms/inference). `CPUAndGPU` gave `Δlogits 6.5e-04`. I'd suggest defaulting to `cpu_and_gpu`, or at least documenting the hazard.
- **`GridSample` under CoreML** returned `max|Δboxes| = 0.945` while logits stayed correct (4.8e-04), even with static shapes. Substituting an equivalent gather+lerp for the deformable-attention `grid_sample` fixed it (I checked that substitution is bit-exact against `F.grid_sample` in isolation), at ~5% CPU cost.

## 4. Whether it pays off depends heavily on the usage pattern

Caveat up front: **these are our numbers for a per-invocation CLI on one machine.** A long-lived server that builds sessions once and serves many requests has completely different economics and may well come out ahead — I haven't measured that case.

With a static+gather export, `MLProgram` + `CPUAndGPU` + the compile cache, raw inference is genuinely faster than CPU. In a tight loop with pre-built sessions I measured 67 pages/s across 6 CoreML sessions vs 17.6 pages/s for an 8-thread CPU pool. **But that synthetic figure did not transfer**: back-solving from end-to-end runs, CoreML achieved roughly half that in situ, presumably because the rest of the pipeline (rendering, resize, OCR) occupies the CPU threads that submit GPU work. I'd treat the 3.8× as an upper bound that our pipeline doesn't reach, not as an expected speedup.

Two things that constrain the design, which I did measure carefully:

- **Session creation is ~2.08 s each and did not parallelize.** Creating N sessions on N threads took the same wall time as creating them sequentially (8 sessions: 16.65 s parallel vs 16.52 s serial, 1.00×). Only the compiled model caches to disk; the session object is rebuilt per process.
- **A single shared session isn't available in Rust.** `ort` takes `&mut self` on `run()` because ONNX Runtime's `Run` isn't thread-safe. (In Python, one session driven by 4+ threads reached 62.9 pages/s at 2.1 s startup — which would change the arithmetic a lot — but that's the pattern `ort` explicitly warns against, so I don't think it counts.)

Fitting our end-to-end runs at three batch sizes (130 / 260 / 520 pages, `w=6`, linear to within 1% across both intervals) gave CPU `0.40 s + 61.1 ms/page` vs CoreML `9.59 s + 42.0 ms/page` — crossover near 480 pages per process. **That fit is specific to this machine, this worker count and this document mix**; I wouldn't generalise the constants. The qualitative shape — a large fixed session cost against a moderate per-page gain — is what I'd expect to hold.

We've dropped CoreML from our tool on that basis, which is a statement about our usage rather than about the EP.

## Suggested resolution

1. Consider not defaulting to `MLProgram` while the shipped layout model aborts under it — or gating that default on a static-shape model being present.
2. Default `MLComputeUnits` to `cpu_and_gpu`, and document that `all` / `cpu_and_ne` can silently corrupt output.
3. Keep `DOCLING_RS_COREML_STATIC_SHAPES` — it's what makes the stock model survive.
4. XNNPACK: agreed with your commit note that no prebuilt ONNX Runtime ships it; opt-in-only seems right.

Separately, this turned up two performance findings that aren't CoreML-specific and look larger than this issue — filed as #338 (layout batching default) and #339 (dynamic batch axis vs NHWC conv kernels).

**Comment by artiz** [comment]
Closed by mistake.
@petri defaults flipped per your findings (static shapes on, cpu_and_gpu). Also I have applied fixes for #338 (layout batching default) and #339 (dynamic batch axis vs NHWC conv kernels). More info are in #325 

Please test it again with docling.rs [1.32.0](https://pypi.org/project/docling-rs/) on your M4 Max and report any errors/performance issues here

**Comment by petri** [comment]
Retested on the M4 Max (12P+4E, macOS 26). Pulled `docling = "1.32"` from crates.io, which resolved to **1.33.0**. Stock `layout_heron.onnx`, no local re-export, CPU provider unless stated. Two documents: 130-page and 113-page industrial manuals. Times are best-of-2 wall clock; output compared byte-for-byte.

**Short version: all three issues look fixed on ARM, and I've deleted the local workarounds this project was carrying.**

## #324 — CoreML defaults: no more abort

With the `coreml` feature compiled, 130-page document:

| `DOCLING_RS_EP` | exit | wall | aborts |
|---|---|---|---|
| unset (→ auto) | 0 | 19.3 s | none |
| `coreml` | 0 | 18.2 s | none |
| `auto` | 0 | 17.3 s | none |
| `cpu` | 0 | 8.3 s | none |

The SIGABRT is gone in every configuration, and CoreML output is **byte-identical** to CPU. The `STATIC_SHAPES`-on + `cpu_and_gpu` defaults do what they needed to.

## Discoverability: `auto`/`coreml` is currently slower here, silently

The one thing I'd push on. With `--features coreml` and `DOCLING_RS_EP` unset, `auto` is **~2.1× slower than CPU** on a one-shot run (17.3 s vs 8.3 s on the 130-page doc). That's your documented residual — session creation is ~2 s per worker and doesn't amortize in a short process — so it's expected behaviour, not a bug. But nothing surfaces it: the run just takes twice as long, with no log line, no warning, and correct output. A user who builds the feature expecting acceleration has no way to discover they've made things slower short of A/B-timing it themselves.

Two things that would help, in increasing order of effort:

1. **Say it in the README** where the `coreml` feature is described — something like "CoreML session creation costs ~2 s per worker and does not parallelize; it pays off for long-lived processes (`docling-serve`) and large batches, and is typically a net loss for one-shot CLI conversions." That alone would have saved me a fair amount of measuring.
2. **A one-time stderr line when CoreML registers**, along the lines of `docling-rs: CoreML registered; session setup ~Ns — for short runs DOCLING_RS_EP=cpu is usually faster`. You already have the once-only warn convention in `ep.rs`, and it would make the tradeoff visible at the moment it's incurred rather than buried in docs.

I'd not suggest changing the `auto` default — the right choice genuinely depends on process lifetime, and you can't know that at session creation. Making the cost legible seems better than guessing on the user's behalf.

## #339 — free-dimension override: confirmed on ARM

This is the number you asked me to check. Comparing the shipped dynamic artifact against the static re-export I had been using locally, everything else held constant (`DOCLING_RS_PDF_WORKERS=10`, CPU EP):

| model | 130-page | 113-page |
|---|---|---|
| stock `layout_heron.onnx` (1.33, free-dim override) | 8.50 s | 7.14 s |
| my static-shape re-export | 8.45 s | 7.08 s |

Within run-to-run noise, and output byte-identical. **The override reproduces the re-export's benefit from the shipped model** — I've removed the local export and the code that swapped it in. That's the outcome I was hoping for; nice that it needed no new artifact.

## #338 — per-EP batch default

Confirmed the CPU default is now per-page (`prefers_fp32()` false → 1). Forcing the old value on the same build still costs what it did before:

| `DOCLING_RS_PDF_LAYOUT_BATCH` | 130-page |
|---|---|
| 1 (new CPU default) | 8.2 s |
| 4 (old default) | 16.3 s |

So the new default is on the right side of it here. I still haven't tested x86 or a GPU provider, so I can't speak to those.

## Net effect downstream

Our tool went from 23.6 s to ~8.2 s on the 130-page document across this whole sequence, with identical output, and now runs **stock docling with no fork and no generated model artifacts**. Thanks for turning these around.

## One observation, take it or leave it

`pdf_worker_count()` caps the pool at 4. On this 16-core machine, raising it to 10 was worth ~1.2× (10.0 s → 8.5 s on the 130-page doc, byte-identical output) — it's the only tuning we still apply locally. I appreciate the cap has a stated rationale (bounded model memory, ~0.4 GB/worker, and memory-bus pressure), and on a 128 GB machine that constraint simply isn't binding, so this may be working as intended for smaller machines. If you ever wanted to make it adaptive rather than fixed, available memory seems like the input — but I have a sample size of one machine and no x86 data, so I'd not change anything on my say-so alone. Happy to file it separately if it's of interest.

**Comment by artiz** [comment]
@petri 
All three of your suggestions from the retest are in (thanks for the numbers — they made every one of these decisions easy):

- README now has a "When CoreML pays off" section right where the coreml feature is described: ~2 s session creation per worker, non-parallelizing, amortizes for long-lived processes (docling-serve) and large batches, typically a net loss for one-shot CLI conversions with the crossover around a few hundred pages per process.
- One-time stderr line at CoreML registration (via the existing once-only convention in the EP apply path): docling-rs: CoreML registered; its session setup costs roughly 2 s per worker and pays off for long-lived processes and large batches — for short one-shot runs DOCLING_RS_EP=cpu is usually faster. The auto default itself is unchanged, per your argument — process lifetime isn't knowable at session creation, so we make the cost legible instead of guessing.
- The worker cap is no longer 4. pdf_worker_count() keeps the `workers ≈ cores / 2` rule but the ceiling moves to 16 and is now explicitly a memory bound (~0.4 GB of model sessions per worker, so a worst-case pool stays under ~6.5 GB even on a ≥32-core host; docling-serve additionally sits behind its `DOCLING_RS_MAX_MEMORY_MB` admission control). On your 16-core machine the default becomes 8 workers — most of the 1.2× you measured at 10 — and `DOCLING_RS_PDF_WORKERS` still overrides both ways, so your =10 keeps working. Machines with ≤4 effective threads keep the exact old sizing. No separate issue needed.


On the version note: docling = "1.32" resolving to 1.33.0 is cargo's standard caret/semver behavior (^1.32), not a packaging quirk on our side — pin with =1.32.0 if you need reproducibility. And auto has been the default whenever a GPU-class feature is compiled in since #74, which is exactly what your "unset (→ auto)" row exercised.

Landing in the next release. Closing with it — and thank you again for measurement quality most PRs can only dream of.
